main.tf

```
@@ -205,34 +218,69 @@ data "aws_ami" "runner" {
   owners = var.ami_owners
 }
 
-resource "aws_launch_configuration" "gitlab_runner_instance" {
-  name_prefix          = var.runners_name
-  security_groups      = [aws_security_group.runner.id]
-  key_name             = var.ssh_key_pair
-  image_id             = data.aws_ami.runner.id
-  user_data            = local.template_user_data
-  instance_type        = var.instance_type
-  ebs_optimized        = var.runner_instance_ebs_optimized
-  enable_monitoring    = var.runner_instance_enable_monitoring
-  spot_price           = var.runner_instance_spot_price
-  iam_instance_profile = aws_iam_instance_profile.instance.name
-  dynamic "root_block_device" {
+resource "aws_launch_template" "gitlab_runner_instance" {
+  name_prefix            = var.runners_name
+  key_name               = var.ssh_key_pair
+  image_id               = data.aws_ami.runner.id
+  user_data              = base64encode(local.template_user_data)
+  instance_type          = var.instance_type
+  update_default_version = true
+  ebs_optimized          = var.runner_instance_ebs_optimized
+  monitoring {
+    enabled = var.runner_instance_enable_monitoring
+  }
+  dynamic "instance_market_options" {
+    for_each = var.runner_instance_spot_price != null && length(var.runner_instance_spot_price) == 0 ? [] : ["spot"]
```

The array should be set with spot once the runner_instance_spot_price is set, same as for tag_specification
```suggestion
    for_each = var.runner_instance_spot_price == null || var.runner_instance_spot_price == "" ? [] : ["spot"]
```

---

Tested with this changes and it looks all good!